numpy/lib/format.py

```diff
@@ -623,13 +623,26 @@ def _read_array_header(fp, version, max_header_size=_MAX_HEADER_SIZE):
     #   "descr" : dtype.descr
     # Versions (2, 0) and (1, 0) could have been created by a Python 2
     # implementation before header filtering was implemented.
-    if version <= (2, 0):
-        header = _filter_header(header)
+    #
+    # For performance reasons, we try without _filter_header first though
     try:
         d = safe_eval(header)
     except SyntaxError as e:
-        msg = "Cannot parse header: {!r}"
-        raise ValueError(msg.format(header)) from e
+        if version <= (2, 0):
+            header = _filter_header(header)
+            try:
+                d = safe_eval(header)
+            except SyntaxError as e2:
+                msg = "Cannot parse header: {!r}"
+                raise ValueError(msg.format(header)) from e2
+            else:
+                warnings.warn(
+                    "Read array created with Python 2, requires header "
+                    "filtering. Consider saving with Python 3 for faster "
+                    "loading.", UserWarning, stacklevel=2)
```

Comment:
On the call nobody had reservations about this, so I guess we keep it.  (I did not think about libraries though, if libraries package `.npy`'s a warning would be pretty annoying.)

I don't have any good ideas, but I feel we should be able to make the text a bit clearer.  Just thinking out loud: ```Reading `.npy` or `.npz` file required additional header parsing as it was created on Python 2.  Save the file again to speed up loading and avoid this warning.```

(I guess `stacklevel=4` or so to show up in `np.load` itself would break other functions, so gotta stick with it.)

Reply:
This was the best I could come up with, your version sounds much better! I'll replace it. Could you elaborate on the `stacklevel`? I don't fully understand.

Reply:
Well, the current stacklevel will point will point one function call up.  Actually, I guess we are at least 4 up (in user space?).  Try:
```
>>> np.load(something_that_warns)
```
And see which line gets reported.  If it is in private NumPy functionality, the stacklevel is set too low.

Reply:
Can this be automatically tested? Does it make sense to add such a test?

Reply:
If you really want to, sure. But unless `pytest.warns` grows it as a feature, I wouldn't bother.  I guess that could be neat.